### ISSUE TITLE
npm 5 (5.0.0 / 5.0.0-beta64) throws on fetching package metadata from a git URL

### ISSUE THREAD

**Issue by mnquintana** [npm/npm#16726]
#### I'm opening this issue because:

  - [x] npm is crashing.
  - [ ] npm is producing an incorrect install.
  - [ ] npm is doing something I don't understand.
  - [ ] Other (_see below for feature requests_):

#### What's going wrong?
When installing a dependency from a git URL, npm 5.0.0 crashes with a git error code (`exited with error code: 128`). This happens even when installing directly from the CLI (as opposed to from `npm-shrinkwrap.json` / `package-lock.json`).

npm 4 can install this same git dependency (although only after completing the 2FA auth via another service first 😅)

#### How can the CLI team reproduce the problem?
Not sure of exact repro steps yet 😞 other than that it involves a git dependency on an instance of GitHub Enterprise that requires ssh auth. TBDDD

Excerpt from `npm-debug.log`:
```
npm WARN read-shrinkwrap This version of npm is compatible with lockfileVersion@1, but npm-shrinkwrap.json was generated for lockfileVersion@0. I'll try to do my best with it!
209 silly fetchPackageMetaData error for analyze-the-thing@git+ssh://git@<GHE URL goes here>.com:<user>/analyze-the-thing.git#fda6f1e2f74f61a9ba80d814f28b8b37a9c3abf7 exited with error code: 128
210 verbose stack Error: exited with error code: 128
210 verbose stack     at ChildProcess.onexit (/usr/local/lib/node_modules/npm5/node_modules/mississippi/node_modules/end-of-stream/index.js:39:36)
210 verbose stack     at emitTwo (events.js:106:13)
210 verbose stack     at ChildProcess.emit (events.js:191:7)
210 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:215:12)
```

Excerpt from `npm-shrinkwrap.json`:
```json
    "analyze-the-thing": {
      "version": "1.0.0",
      "from": "git+ssh://git@<GHE URL goes here>.com:<user>/analyze-the-thing.git#fda6f1e",
      "resolved": "git+ssh://git@<GHE URL goes here>.com:<user>/analyze-the-thing.git#fda6f1e2f74f61a9ba80d814f28b8b37a9c3abf7",
      "dev": true
    },
```

### supporting information:

 - `npm -v` prints: 5.0.0-beta64
 - `node -v` prints: 7.4.0
 - `npm config get registry` prints:
 - Windows, OS X/macOS, or Linux?: macOS 10.12.5

**Comment by jbcpollak** [comment]
Hi have the same problem with `git+https://` dependencies and public Git:

The dependency in `package.json`:

` "loopback-component-bunyan": "git+https:6RiverSystems/loopback-component-bunyan.git#v0.0.6",`

and the error in the log file:

```
29 silly fetchPackageMetaData error for loopback-component-bunyan@git+https:6RiverSystems/loopback-component-bunyan.git#v0.0.6 exited with error code: 128
30 verbose stack Error: exited with error code: 128
30 verbose stack     at ChildProcess.onexit (/Users/jpollak/.nodenv/versions/7.5.0/lib/node_modules/npm/node_modules/mississippi/node_modules/end-of-stream/index.js:39:36)
30 verbose stack     at emitTwo (events.js:106:13)
30 verbose stack     at ChildProcess.emit (events.js:192:7)
30 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:215:12)
```

npm: 5.0.0
node: v7.5.0
MacOS 10.12.5

**Comment by romellem** [comment]
Confirming mostly what @jbcpollak added, but I am seeing it on `git+ssh://` dependencies on private Gitlab instance:

```
1991 silly fetchPackageMetaData error for test@git+ssh://git@gitlab.com:test/test.git exited with error code: 128
1992 verbose stack Error: exited with error code: 128
1992 verbose stack     at ChildProcess.onexit (/usr/local/lib/node_modules/npm/node_modules/mississippi/node_modules/end-of-stream/index.js:39:36)
1992 verbose stack     at emitTwo (events.js:106:13)
1992 verbose stack     at ChildProcess.emit (events.js:194:7)
1992 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:215:12)
```

**Comment by jbcpollak** [comment]
Here is a minimal `package.json` that reproduces it for me:

```
{
  "name": "npm-issue-16726",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "",
  "license": "ISC",
  "dependencies": {
    "loopback-component-bunyan": "git+https:6RiverSystems/loopback-component-bunyan.git#v0.0.6"
  }
}
```

**Comment by jbcpollak** [comment]
update: I realized my dependency was incorrect - it should be as follows, which works correctly:

`"loopback-component-bunyan": "git+https://github.com/6RiverSystems/loopback-component-bunyan.git#v0.0.6"`

I believe the problem @mnquintana and @romellem has the same symptoms but a different root cause. In my case it would be helpful to have better diagnostic information on the console, but when the package is specified correctly it seems to work fine.

**Comment by ds82** [comment]
I'm seeing the same problem/error with a private gitlab (self-hosted) git URL. Works with npm@3 and npm@4.

```
~/c/t/titan-add-on-univers ❯❯❯ npm i git+ssh://git@rj-gitlab.rea.de:web/node-xml2js.git#07be3dd1af6dfb836d29290c3dba78cc9239add1                                                                                              build/mbh/uni ◼
npm WARN read-shrinkwrap This version of npm is compatible with lockfileVersion@1, but npm-shrinkwrap.json was generated for lockfileVersion@0. I'll try to do my best with it!
npm ERR! exited with error code: 128

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/dsaenger/.npm/_logs/2017-05-29T06_10_32_619Z-debug.log
~/c/t/titan-add-on-univers ❯❯❯ cat /Users/dsaenger/.npm/_logs/2017-05-29T06_10_32_619Z-debug.log                                                                                                                            ⏎ build/mbh/uni ◼
0 info it worked if it ends with ok
1 verbose cli [ '/Users/dsaenger/.nvm/versions/node/v4.8.3/bin/node',
1 verbose cli   '/Users/dsaenger/.nvm/versions/node/v4.8.3/bin/npm',
1 verbose cli   'i',
1 verbose cli   'git+ssh://git@rj-gitlab.rea.de:web/node-xml2js.git#07be3dd1af6dfb836d29290c3dba78cc9239add1' ]
2 info using npm@5.0.0
3 info using node@v4.8.3
4 silly install loadCurrentTree
5 silly install readLocalPackageData
6 warn read-shrinkwrap This version of npm is compatible with lockfileVersion@1, but npm-shrinkwrap.json was generated for lockfileVersion@0. I'll try to do my best with it!
7 silly fetchPackageMetaData error for git+ssh://git@rj-gitlab.rea.de:web/node-xml2js.git#07be3dd1af6dfb836d29290c3dba78cc9239add1 exited with error code: 128
8 verbose stack Error: exited with error code: 128
8 verbose stack     at ChildProcess.onexit (/Users/dsaenger/.nvm/versions/node/v4.8.3/lib/node_modules/npm/node_modules/mississippi/node_modules/end-of-stream/index.js:39:36)
8 verbose stack     at emitTwo (events.js:87:13)
8 verbose stack     at ChildProcess.emit (events.js:172:7)
8 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:211:12)
9 verbose cwd /Users/dsaenger/code/titan-addons/titan-add-on-univers
10 verbose Darwin 14.5.0
11 verbose argv "/Users/dsaenger/.nvm/versions/node/v4.8.3/bin/node" "/Users/dsaenger/.nvm/versions/node/v4.8.3/bin/npm" "i" "git+ssh://git@rj-gitlab.rea.de:web/node-xml2js.git#07be3dd1af6dfb836d29290c3dba78cc9239add1"
12 verbose node v4.8.3
13 verbose npm  v5.0.0
14 error exited with error code: 128
15 verbose exit [ 1, true ]
```

**Comment by zkat** [comment]
This should be fixed once https://github.com/npm/npm-package-arg/pull/24 lands.

I'll be closing all the other issues related to this as duplicates and keep this one as the main one.

I believe all the _other_ git identifier issues have been resolved, and all supported git URIs should be working again, as well as `#semver:` range specs.

The issue basically comes down to `git@foo.com:username/project.git` not being an actual URI. It is, in fact, a bullshit specifier exclusively from `scp`, but it's pretty handy, particularly when self-hosting git projects. It's _technically_ incorrect for us to be doing this thing where we support them as `git+ssh:` specs (because they're not URIs), but we're stuck with sins of the past. We have special-case code for this in `hosted-git-info`, and the rest of it used to live in npm's own git handling (it was a _mess_), so it just kinda disappeared during the rewrite. Oops lol

**Comment by zkat** [comment]
This has been fixed and merged into release-next. It will be available with 5.0.2 when it's released (maybe tomorrow?).

It is also available _right now_ by using [the canary version of npm](https://npm.im/npmc), which you can just install with `npm i -g npmc`. I'll wait to close this until I get confirmation from y'all that this does what I believe it does ^_^;; (it did work when I tested it myself on a non-hosted git)

**Comment by AleBles** [comment]
Had the same issue as @romellem , can confirm that the issue is not there anymore on NPM canary. Thanks @zkat !

**Comment by swist** [comment]
I'm still having the same problem with npm 5.0.1-canary.6, installed it by running
```npm install -g npmc``` 


I'm trying to install a lib off our internal bitbucket repo

**Comment by chrispappas** [comment]
```
npm install git+ssh://git@<private-gitlab-somewhere>/<group>/<repo>.git
npm ERR! cb() never called!

npm ERR! This is an error with npm itself. Please report this error at:
npm ERR!     <https://github.com/npm/npm/issues>
```
```
0 info it worked if it ends with ok
1 verbose cli [ '/usr/local/Cellar/node/8.0.0_1/bin/node',
1 verbose cli   '/usr/local/bin/npm',
1 verbose cli   'install',
1 verbose cli   'git+ssh://git@<private-gitlab-somewhere>/<group>/<repo>.git' ]
2 info using npm@5.0.1
3 info using node@v8.0.0
4 silly install loadCurrentTree
5 silly install readLocalPackageData
6 error cb() never called!
7 error This is an error with npm itself. Please report this error at:
8 error <https://github.com/npm/npm/issues>
```

**Comment by zkat** [comment]
@swist do you have a debug log if this is still happening?

**Comment by chuckhacker** [comment]
Still seeing this issue in npm 5.0.2

**Comment by zkat** [comment]
@chuckhacker can you paste the debug log with the failing URI?

**Comment by jonathantneal** [comment]
I had this issue as well. It has been resolved since I switched to 5.0.3 using `npm install -g npm@next`.

My log was (with minor alterations):

```
1517 silly fetchPackageMetaData error for my-package@git+ssh://git@my-domain:repo.git#tag@1.0.0 exited with error code: 128
1518 verbose stack Error: exited with error code: 128
1518 verbose stack     at ChildProcess.onexit (/usr/local/lib/node_modules/npm/node_modules/mississippi/node_modules/end-of-stream/index.js:39:36)
1518 verbose stack     at emitTwo (events.js:125:13)
1518 verbose stack     at ChildProcess.emit (events.js:213:7)
1518 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:197:12)
1519 verbose cwd /Users/me/my-component
1520 verbose Darwin 16.6.0
1521 verbose argv "/usr/local/Cellar/node/8.0.0_1/bin/node" "/usr/local/bin/npm" "install"
1522 verbose node v8.0.0
1523 verbose npm  v5.0.1
1524 error exited with error code: 128
1525 verbose exit [ 1, true ]
```

**Comment by stridera** [comment]
I'm still getting the problem using 5.0.3
```
02 silly fetchPackageMetaData error for repo@git+ssh://git@gitenterprise.domain.com:user/repo.git exited with error code: 128
503 verbose stack Error: exited with error code: 128
503 verbose stack     at ChildProcess.onexit (/Users/user/.nvm/versions/node/v4.8.3/lib/node_modules/npm/node_modules/mississippi/node_modules/end-of-stream/index.js:39:36)
503 verbose stack     at emitTwo (events.js:87:13)
503 verbose stack     at ChildProcess.emit (events.js:172:7)
503 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:211:12)
504 verbose cwd /Users/user/code/repox	
505 verbose Darwin 16.6.0
506 verbose argv "/Users/user/.nvm/versions/node/v4.8.3/bin/node" "/Users/user/.nvm/versions/node/v4.8.3/bin/npm" "link"
507 verbose node v4.8.3
508 verbose npm  v5.0.3
509 error exited with error code: 128
510 verbose exit [ 1, true ]

**Comment by amertak** [comment]
Still not resolved, looking for '#latest' even with '#master' specified.

**Comment by zkat** [comment]
@amertak can you reproduce this in the canary that has the latest patches? (install it with `npm i -g npmc@latest` and run it with `npmc`).

**Comment by amertak** [comment]
@zkat It is not reproducible with the latest npmc (5.0.3-canary.12). It works correctly.

Can I expect the fix in the 5.0.4?

Thanks ;)

**Comment by jonathantneal** [comment]
Tested `npm install --global npmc@latest`:

```
npmc -v
5.0.3-canary.12
```

Followed that with `npm install`, and all went well. But then I tried `npm install --production`:

```
removed 947 packages in 13.31s
```

So, this is still an issue for me.

**Comment by amertak** [comment]
@jonathantneal You have to run `npmc install`

**Comment by jonathantneal** [comment]
@amertak, my apologies, I was using npmc, so here’s a bigger output.

After `npmc install` I do:

```
npmc install --production
npm notice CANARY npmc is experimental software. If you find an issue, please file it in the main npm repository, and call out that you were using npmc.

> pre-commit@1.2.2 uninstall /path/to/node_modules/pre-commit
> node uninstall.js

removed 947 packages in 13.095s
```

**Comment by amertak** [comment]
So I have an update.

One github user, two repos. One works, second does not. 
Says it cannot find it. Weird.

**Comment by pyskell** [comment]
Adding to this. I get the same issue and error with `npm 5.0.0` through `npmc 5.0.3-canary.12`. I'm on `Ubuntu 16.04 LTS`. Error is exactly as others have already described.

**Comment by shane-walker** [comment]
I'm seeing this issue too. I was on `npm 5.0.3` and I tried the canary but had to roll back to `4.x`.

I have two different git sources I'm pulling from with 9 different repos between them.

**Comment by pyskell** [comment]
`npm 4.6.1` does not have this issue. Everything from `npm 5.0.0` onward does.

**Comment by amertak** [comment]
Can we please reopen this issue?

**Comment by alexfernandez** [comment]
Running 5.0.3, still broken here. Package URL is: `git+ssh://user@selfhosting.io:repo`. Please reopen.

**Comment by vforvalerio87** [comment]
I can confirm that this does not work in any version of npm 5; we are still using npm 4.6.1 to make it work

**Comment by alexfernandez** [comment]
Works with `npmc` here:

```
 $ npm i -g npmc
 $ npmc install git+ssh://user@selfhosting.io:repo
npm notice CANARY npmc is experimental software. If you find an issue, please file it in the main npm repository, and call out that you were using npmc.
+ repo@1.5.3
added 37 packages and removed 1 package in 9.286s
```

`npmc install` did not however install the self-hosted package if missing, but one thing at a time.

I understand `npmc` should be released as 5.0.4, soon 🤞  [Policy is](https://twitter.com/izs/status/877657885831962624) that issues fixed in `npmc` are not reopened.

**Comment by zspitz** [comment]
I'm seeing this issue on both `npm` and `npmc`, trying to run `npm install` on a clone of https://github.com/DefinitelyTyped/DefinitelyTyped.

Full logs are at DefinitelyTyped/DefinitelyTyped#17491.

npm:
```
npm ERR! exited with error code: 128

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\win8\AppData\Roaming\npm-cache\_logs\2017-06-26T13_12_02_466Z-debug.log
```
npmc:
```
npm notice CANARY npmc is experimental software. If you find an issue, please file it in the main npm repository, and call out that you were using npmc.
npm ERR! Error while executing:
npm ERR! C:\Program Files\Git\cmd\git.EXE ls-remote -h -t git://github.com/microsoft/definitelytyped-header-parser.git
npm ERR!
npm ERR! fatal: Could not read from remote repository.
npm ERR!
npm ERR! Please make sure you have the correct access rights
npm ERR! and the repository exists.
npm ERR!
npm ERR! exited with error code: 128

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\win8\AppData\Roaming\npm-cache\_logs\2017-06-26T13_17_57_917Z-debug.log
```
I can successfully clone from `git://github.com/microsoft/definitelytyped-header-parser.git`.

**Comment by yanzhihong23** [comment]
Still seeing this issue on `npm 5.5.1`.

**Comment by vforvalerio87** [comment]
I can confirm this still does not working with codecommit urls; we are still stuck with npm 4.6.1
